pkg/sources/udsource/grpc_udsource.go

```diff
@@ -0,0 +1,161 @@
+				return nil
+			}
+			time.Sleep(1 * time.Second)
+		}
+	}
+}
+
+// ApplyPendingFn returns the number of pending messages in the source.
+func (u *GRPCBasedUDSource) ApplyPendingFn(ctx context.Context) (int64, error) {
+	if resp, err := u.client.PendingFn(ctx, &emptypb.Empty{}); err == nil {
+		return int64(resp.Result.Count), nil
+	} else {
+		return isb.PendingNotAvailable, err
+	}
+}
+
+// ApplyReadFn reads messages from the source.
+func (u *GRPCBasedUDSource) ApplyReadFn(ctx context.Context, count int64, timeout time.Duration) ([]*isb.ReadMessage, error) {
+	var readMessages []*isb.ReadMessage
+
+	// Construct the gRPC request
+	var r = &sourcepb.ReadRequest{
+		Request: &sourcepb.ReadRequest_Request{
+			NumRecords:  uint64(count),
+			TimeoutInMs: uint32(timeout.Milliseconds()),
+		},
+	}
+
+	// Prepare the ReadResponse channel
+	var datumCh = make(chan *sourcepb.ReadResponse)
+	// Prepare the error channel to receive errors from the ReadFn goroutine
+	errCh := make(chan error, 1)
+	defer close(errCh)
+	var wg sync.WaitGroup
+	wg.Add(1)
+
+	// Start the goroutine to read messages and send to the channel
+	go func() {
+		defer wg.Done()
+		defer close(datumCh)
+		if err := u.client.ReadFn(ctx, r, datumCh); err != nil {
+			errCh <- fmt.Errorf("failed to read messages from udsource: %w", err)
+		}
+	}()
+
+	// Collect the messages from the channel and return
+	for {
+		select {
+		case <-ctx.Done():
+			// If the context is done, return the messages collected so far
+			return readMessages, fmt.Errorf("context is done, %w", ctx.Err())
+		case err := <-errCh:
+			// If the ReadFn goroutine returns an error, return the messages collected so far
+			return readMessages, err
+		case datum, ok := <-datumCh:
+			if !ok {
+				// If the channel is closed, wait for the ReadFn goroutine to finish
+				wg.Wait()
+				return readMessages, nil
+			}
+			// Convert the datum to ReadMessage and append to the list
+			r := datum.GetResult()
+			readMessage := &isb.ReadMessage{
+				Message: isb.Message{
+					Header: isb.Header{
+						MessageInfo: isb.MessageInfo{EventTime: r.GetEventTime().AsTime()},
+						ID:          constructMessageID(r),
+						Keys:        r.GetKeys(),
+					},
+					Body: isb.Body{
+						Payload: r.GetPayload(),
+					},
+				},
+				ReadOffset: utils.ConvertToIsbOffset(r.GetOffset()),
+			}
+			readMessages = append(readMessages, readMessage)
+		}
+	}
+}
+
+// ApplyAckFn acknowledges messages in the source.
+func (u *GRPCBasedUDSource) ApplyAckFn(ctx context.Context, offsets []isb.Offset) error {
+	rOffsets := make([]*sourcepb.Offset, len(offsets))
+	for i, offset := range offsets {
+		rOffsets[i] = utils.ConvertToSourceOffset(offset)
+	}
+	var r = &sourcepb.AckRequest{
+		Request: &sourcepb.AckRequest_Request{
+			Offsets: rOffsets,
+		},
+	}
+	_, err := u.client.AckFn(ctx, r)
+	return err
+}
+
+func constructMessageID(r *sourcepb.ReadResponse_Result) string {
```

Comment:
make it a stringer?

Reply:
The proto generated `sourcepb.ReadResponse_Result` does have a stringer which uses the following method to translate the message to a string.

```
// MessageStringOf returns the message value as a string,
// which is the message serialized in the protobuf text format.
func (Export) MessageStringOf(m protoreflect.ProtoMessage) string {
	return prototext.MarshalOptions{Multiline: false}.Format(m)
}
```

I don't feel comfortable using it because it's not within our control. It might not enforce uniqueness and it the output can be quite verbose, probably the ID generated by it is longer than my current approach.